### ISSUE TITLE
List addition implementation

### ISSUE THREAD

**Issue by ThakeeNathees** [ThakeeNathees/pocketlang#55]
## [TODO: list addition](https://github.com/ThakeeNathees/pocketlang/blob/749456e215590218afa6dead42c1259007728ea3/src/pk_core.c#L1008-L1014)

Currently adding 2 lists (`[1, 2] + [3, 4]`) cause a crash with a `TODO` message. which need to be implemented. ([here](https://github.com/ThakeeNathees/pocketlang/blob/749456e215590218afa6dead42c1259007728ea3/src/pk_core.c#L1008-L1014))

## How to implement
`List* l1, *l2` are defined above the `TODO` statement in the source.
- Create a new list using `newList(vm, size)` function, where (initial reserved) size is the total length of l1 and l2.
  length of l1 = `l1->elements.count`
- iterate each lists and append to the new list.
```c
for (uint32_t i = 0; i < l1->elements.count; i++) {
  listAppend(vm, list, l1->elements.data[i]);
}
```
- return the list as a var using `VAR_OBJ(list)` macro function.

## Finally

Once you successfully implemented it, write some tests at `tests/lang/`

**Comment by alexcpatel** [comment]
@ThakeeNathees I can take this one later tonight :). Unless you’d like to keep it open

**Comment by ThakeeNathees** [comment]
@alexcpatel that's great, I'll close this issue once your PR is merged (like everyother).

Here is something you need to consider when implementing.
- To get the initial length for the new list `l1->elements.count + l2->elements.count`

- Also you need to push the new list to the VM's root (where all the temporary objects live) so that any allocations (that'll trigger garbage collection) won't `free()` the list.

```c
List* list = newList(vm, size);
vmPushTempRef(vm, &list->_super); // protect list from garbage collected

// iterate through List l1 and add their elements.
// iterate through List l2 and add their elements.

vmPopTempRef(vm); // release the list from the vm's root

return VAR_OBJ(list);
```

I'll soon make an architecture overview, explaining how pocketlang internal works and the garbage collection mechanism.

**Comment by alexcpatel** [comment]
@ThakeeNathees I do need to learn more about the garbage collection system (a doc would be really helpful, although eventually I will dive into the code which may take a while).

Is it not possible for the `list` result from `newList()` to be freed before the `vmPushTempRef()` immediately following? This can crash the VM?

**Comment by ThakeeNathees** [comment]
> Is it not possible for the list result from newList() to be freed before the vmPushTempRef() immediately following? This can crash the VM?

No, A garbage collection is only triggered when the pocketlang calls allocation ([see here](https://github.com/ThakeeNathees/pocketlang/blob/5f6f66517fbc3779e3db7960ab05d8fa76246a3f/src/pk_vm.c#L212-L226)). Appending elements to the list may invoke an allocation call if the list is filled (this may help you #78), so we need the list to be protected while we're appending elements to it.

I'll create [reference](https://github.com/ThakeeNathees/pocketlang/issues?q=label%3Areference+) labeled issue explaining how the garbage collection works soon, and which could be used to ask questions or discuss suggestions and improvements.

**Comment by alexcpatel** [comment]
Ah, for some reason I thought some of this was multithreaded (with all the fiber stuff). Makes sense, thanks!

**Comment by ThakeeNathees** [comment]
Fixed by #80

**Comment by ThakeeNathees** [comment]
> @ThakeeNathees I do need to learn more about the garbage collection system (a doc would be really helpful, although eventually I will dive into the code which may take a while).

I've been writing internal docs for some time (not good at writing docs thought), just finished the GC today, here: #88 hope this will help you understand.

@alexcpatel
